### PR TITLE
Fix reorder policy job to skip compressed chunks

### LINKED ISSUE #1810: Reorder job fails because of already compressed chunks
**Relevant system information:**

 PostgreSQL 11.6 (Debian 11.6-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
 - TimescaleDB version: 1.6.0
 - Deployed using TimescaleDB Helm charts

**Describe the bug**

We use the following compression policy:

```
ALTER TABLE timeseries.quantity SET (timescaledb.compress, timescaledb.compress_orderby = 'event_time, sequence', timescaledb.compress_segmentby = 'metric_ref');
```

for this table

```
create table quantity
(
	metric_ref integer not null,
	event_time timestamp not null,
	sequence integer not null,
	value double precision
);

create unique index quantity_pk
	on quantity (metric_ref asc, event_time desc, sequence desc);
```

Now we have added a 

```
SELECT add_reorder_policy('timeseries.quantity', 'quantity_pk');
```

but the job fails with

```
2020-04-10 16:47:56 UTC | 29646 | 5e90a33c.73ce-2 | @ | 22023 | ERROR:  "_hyper_3_5237_chunk" is a compressed chunk
2020-04-10 16:47:56 UTC | 64 | 5e70f65c.40-103 | @ | 00000 | LOG:  background worker "Reorder Background Job" (PID 29646) exited with exit code 1
```

**Expected behavior**

We expected the job to finish by skipping already compressed chunks (as they cannot be changed and are already ordered).

### PR DESCRIPTION
**Pull request by Olernov**
Reorder policy does not skip compressed chunks when selecting
next chunk to apply reordering. This causes an error in job
execution since it's not possible to reorder compressed chunks.
With this fix the job excludes compressed chunks from selection.

Fixes #1810

### CODE REVIEW

**Comment by Olernov** [comment]
I can't figure out why my new test fails in Travis. What I did is I renamed the existing test tsl/test/sql/compress_bgw_drop_chunks.sql to compress_bgw_reorder_drop_chunks.sql added new logic and replaced expected/*.out file correspondingly. I ran the test suite against PG 11.7 and 12.2 and they succeeded on my machine, but Travis displays that test _compress_bgw_reorder_drop_chunks_ fails.

File: tsl/test/sql/CMakeLists.txt [hunk 1]
```diff
@@ -58,7 +58,7 @@ if (${PG_VERSION_MAJOR} GREATER "9")
     compression_hypertable.sql
     compression_segment_meta.sql
     compression_bgw.sql
-    compress_bgw_drop_chunks.sql
+    compress_bgw_reorder_drop_chunks.sql
```
Comment (k-rus) [inline tsl/test/sql/CMakeLists.txt:61]: The added test doesn't fit well into this test file, since the file is about compression background worker, while the test is about reorder background worker.

Comment (k-rus) [reply tsl/test/sql/CMakeLists.txt:61]: Will it be more natural to add the test to `bgw_reorder_drop_chunks.sql` and rename it to be for compressed and drop chunks?

**Comment by k-rus** [comment]
> After the fix there still might be a problem when compress bgw and reorder bgw, which will try to compress and reorder the same chunk simultaneously. However, I think it is a separate issue and PR.

After this fix it might be more likely that the collision will happen, while before the fix the reorder job fails and doesn't run. May be a better fix will need to take in account compress jobs.

File: tsl/test/sql/CMakeLists.txt (hunk 1 above)
Comment (Olernov) [reply tsl/test/sql/CMakeLists.txt:61]: But `bgw_reorder_drop_chunks.sql` runs for all versions of PG while `compress_bgw_reorder_drop_chunks.sql` runs only for version higher than 9. If I add compression tests to `bgw_reorder_drop_chunks.sql` it will fail for PG 9.6. So I added them to `compress_bgw_reorder_drop_chunks.sql`.

Comment (k-rus) [reply tsl/test/sql/CMakeLists.txt:61]: You are right.

Comment (k-rus) [reply tsl/test/sql/CMakeLists.txt:61]: I can think to improve the name of this test: `bgw_compress_reorder_drop_chunks.sql`. However, I don't know how others will like this idea :)

**Comment by Olernov** [comment]
> > After the fix there still might be a problem when compress bgw and reorder bgw, which will try to compress and reorder the same chunk simultaneously. However, I think it is a separate issue and PR.
> 
> After this fix it might be more likely that the collision will happen, while before the fix the reorder job fails and doesn't run. May be a better fix will need to take in account compress jobs.

Won't StartTransactionCommand() at job.c help?
```
bool
execute_reorder_policy(BgwJob *job, reorder_func reorder, bool fast_continue)
{
	...
	if (!IsTransactionOrTransactionBlock())
	{
		started = true;
		StartTransactionCommand();
	}
```

I think you talk about a more general problem than bgw jobs. Not only jobs, but users may also call compress_chunk() and reorder_chunk() on a same chunk simultaneously. I believe some kind of exclusive locking must be implemented already.

File: tsl/test/sql/CMakeLists.txt (hunk 1 above)
Comment (Olernov) [reply tsl/test/sql/CMakeLists.txt:61]: No problem, if you prefer another name or a separated file for this test - let me know.

File: src/dimension_slice.c [hunk 2]
```diff
@@ -793,7 +795,7 @@ dimension_slice_check_is_chunk_uncompressed_tuple_found(TupleInfo *ti, void *dat
 	foreach (lc, chunk_ids)
 	{
 		int32 chunk_id = lfirst_int(lc);
-		if (ts_chunk_can_be_compressed(chunk_id))
+		if (ts_chunk_can_be_compressed_or_reordered(chunk_id))
```
Comment (gayyappan) [inline src/dimension_slice.c:798]: Shouldn't combine the 2 different policies together. This is only for compress chunk job. Need functionality specifically for reorder chunks. (similar to ts_dimension_slice_get_chunkid_to_compress )

Comment (Olernov) [reply src/dimension_slice.c:798]: Actually it's not a combination of policies, it's just a check shared among both policies. Both policies require the chunk not to be compressed or dropped. I don't think it makes sense to clone `ts_chunk_can_be_compressed `into `ts_chunk_can_be_reordered` 'cause the code will be duplicated. Maybe rename this check to `ts_chunk_is_not_compressed_or_dropped`?

File: src/chunk.c [hunk 3]
```diff
@@ -2812,10 +2812,10 @@ ts_chunks_in(PG_FUNCTION_ARGS)
 	pg_unreachable();
 }
 
-/* Check if this chunk can be compressed, that it is not dropped and has not
- * already been compressed. */
+/* Check if this chunk can be compressed or reordered, that it is not dropped
+ * and has not already been compressed. */
 bool
-ts_chunk_can_be_compressed(int32 chunk_id)
+ts_chunk_can_be_compressed_or_reordered(int32 chunk_id)
```
Comment (gayyappan) [inline src/chunk.c:2818]: could you revert this change ?

File: src/dimension_slice.c [hunk 4]
```diff
@@ -744,13 +744,14 @@ dimension_slice_check_chunk_stats_tuple_found(TupleInfo *ti, void *data)
 
 	foreach (lc, chunk_ids)
 	{
-		BgwPolicyChunkStats *chunk_stat =
-			ts_bgw_policy_chunk_stats_find(info->job_id, lfirst_int(lc));
+		int chunk_id = lfirst_int(lc);
+		BgwPolicyChunkStats *chunk_stat = ts_bgw_policy_chunk_stats_find(info->job_id, chunk_id);
```
Comment (gayyappan) [inline src/dimension_slice.c:749]: Could you add a comment here
```suggestion
Look for a chunk that a) doesn't have a job stat (reorder ) and b) is not compressed (should not reorder compressed chunk)
```

File: src/dimension_slice.c [hunk 5]
```diff
@@ -759,10 +760,11 @@ dimension_slice_check_chunk_stats_tuple_found(TupleInfo *ti, void *data)
 }
 
 int
-ts_dimension_slice_oldest_chunk_without_executed_job(int32 job_id, int32 dimension_id,
-													 StrategyNumber start_strategy,
-													 int64 start_value, StrategyNumber end_strategy,
-													 int64 end_value)
+ts_dimension_slice_oldest_valid_chunk_for_job_execution(int32 job_id, int32 dimension_id,
```
Comment (gayyappan) [inline src/dimension_slice.c:763]: This is the reorder specific function. Lets rename this.
```suggestion
ts_dimension_slice_oldest_valid_chunk_for_reorder(int32 job_id, int32 dimension_id,
```

File: src/dimension_slice.c (hunk 2 above)
Comment (gayyappan) [reply src/dimension_slice.c:798]: Could you revert this change?

File: tsl/test/sql/compress_bgw_reorder_drop_chunks.sql [hunk 6]
```diff
@@ -96,3 +96,87 @@ FROM _timescaledb_catalog.chunk chunk
...
+SELECT job_id, last_run_success, total_runs, total_successes, total_failures, total_crashes
+    FROM _timescaledb_internal.bgw_job_stat
+    where job_id=:reorder_job_id;
+
+-- two chunks clustered
+SELECT indexrelid::regclass, indisclustered
+    FROM pg_index
+    WHERE indisclustered = true ORDER BY 1;
+
+-- third call to scheduler: nothing to reorder since an eligible chunk
```
Comment (gayyappan) [inline tsl/test/sql/compress_bgw_reorder_drop_chunks.sql:169]: Don't think we need these additional tests. Can be removed.

File: tsl/test/sql/compress_bgw_reorder_drop_chunks.sql [hunk 7]
```diff
@@ -96,3 +96,87 @@ FROM _timescaledb_catalog.chunk chunk
...
 WHERE uncomp_hyper.table_name like 'test_drop_chunks_table';
+
+------------------------------
+-- Test reorder policy runs on compressed tables. Reorder policy job must skip compressed chunks 
+-- (see issue https://github.com/timescale/timescaledb/issues/1810).
+-- More tests for reorder policy can be found at bgw_reorder_drop_chunks.sql
+------------------------------
+
+CREATE TABLE test_reorder_chunks_table(time int not null, chunk_id int);
+CREATE INDEX ON test_reorder_chunks_table(time);
```
Comment (gayyappan) [inline tsl/test/sql/compress_bgw_reorder_drop_chunks.sql:107]: could you add the name of the index here.

File: tsl/test/sql/compress_bgw_reorder_drop_chunks.sql [hunk 8]
```diff
@@ -96,3 +96,87 @@ FROM _timescaledb_catalog.chunk chunk
...
+    WHERE indisclustered = true ORDER BY 1;
+
+-- second call to scheduler
+SELECT ts_bgw_db_scheduler_test_run_and_wait_for_scheduler_finish(25);
+
+SELECT job_id, last_run_success, total_runs, total_successes, total_failures, total_crashes
+    FROM _timescaledb_internal.bgw_job_stat
+    where job_id=:reorder_job_id;
+
+-- two chunks clustered
```
Comment (gayyappan) [inline tsl/test/sql/compress_bgw_reorder_drop_chunks.sql:164]:
```suggestion
-- two chunks clustered , skips the compressed chunks
```

File: src/dimension_slice.c (hunk 2 above)
Comment (Olernov) [reply src/dimension_slice.c]: Is it OK to call `ts_chunk_can_be_compressed` from `dimension_slice_check_chunk_stats_tuple_found` which is used only for reorder policy? Or I should create a separate function `ts_chunk_can_be_reordered`?

Comment (gayyappan) [reply src/dimension_slice.c]: I don't think we need another function since this is just a check for uncompressed chunk. (The other option is to rename ts_chunk_can_be_compressed as you did. But would prefer to keep the changes minimal.)

File: src/chunk.c [hunk 9]
```diff
@@ -3195,8 +3195,8 @@ ts_chunks_in(PG_FUNCTION_ARGS)
 	pg_unreachable();
 }
 
-/* Check if this chunk can be compressed, that it is not dropped and has not
- * already been compressed. */
+/* Check if this chunk can be compressed or reordered, that it is not dropped
```
Comment (gayyappan) [inline src/chunk.c:3198]: could you revert the comment change as well? We are only checking if it can be compressed
